File: backend/seed.py
import sqlite3
import uuid
import random
from datetime import datetime, timedelta

DB_PATH = "fraud_detection.db"
NUM_TX = 40

channels = ["Online", "ATM", "Branch", "Mobile"]
locations = ["New York", "London", "San Francisco", "Tokyo", "Berlin", "Dubai"]
occupations = ["Engineer", "Doctor", "Student", "Retired", "Teacher", "Other"]
types = ["Debit", "Credit", "Transfer"]

def seed_data():
    conn = sqlite3.connect(DB_PATH)
    c = conn.cursor()
    
    # Check if table exists
    c.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='transactions';")
    if not c.fetchone():
        print("transactions table does not exist. Please run backend first.")
        return

    now = datetime.utcnow()
    
    for i in range(NUM_TX):
        is_fraud = random.random() < 0.15 # 15% fraud
        
        tx_id = str(uuid.uuid4())
        acc_id = f"ACC{random.randint(100, 999)}"
        amount = round(random.uniform(5.0, 5000.0), 2)
        
        if is_fraud:
            amount = round(random.uniform(2000.0, 15000.0), 2)
            
        tx_type = random.choice(types)
        loc = random.choice(locations)
        channel = random.choice(channels)
        merchant = f"Merchant_{random.randint(1, 50)}"
        age = random.randint(18, 80)
        occ = random.choice(occupations)
        bal = round(random.uniform(100.0, 50000.0), 2)
        dur = random.randint(10, 600)
        login = random.randint(1, 5)
        
        if is_fraud:
            login = random.randint(3, 10)
            
        # Distribute over last 7 days
        days_ago = random.uniform(0, 7)
        tx_date = now - timedelta(days=days_ago)
        
        prob = random.uniform(0.6, 0.99) if is_fraud else random.uniform(0.01, 0.3)
        risk = "HIGH" if prob >= 0.75 else "MEDIUM" if prob >= 0.4 else "LOW"
        
        c.execute('''
            INSERT INTO transactions (
                transaction_id, account_id, amount, transaction_type, location, channel, 
                merchant_id, customer_age, customer_occupation, account_balance, 
                transaction_duration, login_attempts, transaction_date, 
                is_fraud, fraud_probability, risk_level, is_reviewed, is_resolved, created_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (
            tx_id, acc_id, amount, tx_type, loc, channel, merchant, age, occ, bal,
            dur, login, tx_date.isoformat(),
            is_fraud, prob, risk, False, False, tx_date.isoformat()
        ))
        
    conn.commit()
    conn.close()
    print(f"Successfully seeded {NUM_TX} transactions!")

if __name__ == "__main__":
    seed_data()

File: frontend/seed.js
import sqlite3 from 'sqlite3';
import path from 'path';
import { fileURLToPath } from 'url';
import { v4 as uuidv4 } from 'uuid';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const dbPath = path.resolve(__dirname, '../backend/fraud_detection.db');
const db = new (sqlite3.verbose()).Database(dbPath);

const NUM_TX = 40;
const channels = ["Online", "ATM", "Branch", "Mobile"];
const locations = ["New York", "London", "San Francisco", "Tokyo", "Berlin", "Dubai"];
const occupations = ["Engineer", "Doctor", "Student", "Retired", "Teacher", "Other"];
const types = ["Debit", "Credit", "Transfer"];

console.log("Connecting to:", dbPath);

db.serialize(() => {
  db.get("SELECT name FROM sqlite_master WHERE type='table' AND name='transactions'", (err, row) => {
    if (!row) {
      console.log("transactions table does not exist. Please run backend first to create tables.");
      return;
    }
    
    db.run("BEGIN TRANSACTION");
    const stmt = db.prepare(`
      INSERT INTO transactions (
        transaction_id, account_id, amount, transaction_type, location, channel, 
        merchant_id, customer_age, customer_occupation, account_balance, 
        transaction_duration, login_attempts, transaction_date, 
        is_fraud, fraud_probability, risk_level, is_reviewed, is_resolved, created_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    let inserted = 0;
    for (let i = 0; i < NUM_TX; i++) {
        const isFraud = Math.random() < 0.15; // roughly 15% fraud
        const txId = uuidv4();
        const accId = `ACC${Math.floor(Math.random() * 900) + 100}`;
        let amount = (Math.random() * 4995) + 5;
        if (isFraud) amount = (Math.random() * 13000) + 2000;
        
        const loc = locations[Math.floor(Math.random() * locations.length)];
        const channel = channels[Math.floor(Math.random() * channels.length)];
        const type = types[Math.floor(Math.random() * types.length)];
        const merchant = `Merchant_${Math.floor(Math.random() * 50) + 1}`;
        const occ = occupations[Math.floor(Math.random() * occupations.length)];
        
        // Random date within last 7 days
        const date = new Date(Date.now() - Math.random() * 7 * 24 * 60 * 60 * 1000);
        
        // Ensure values align
        const prob = isFraud ? (0.6 + Math.random() * 0.39) : (0.01 + Math.random() * 0.29);
        const risk = prob >= 0.75 ? 'HIGH' : prob >= 0.4 ? 'MEDIUM' : 'LOW';

        stmt.run([
            txId, accId, parseFloat(amount.toFixed(2)), type, loc, channel,
            merchant, 18 + Math.floor(Math.random()*60), occ, 1000 + Math.random()*20000,
            10 + Math.floor(Math.random()*200), isFraud ? 5 : 1, date.toISOString().replace('Z',''),
            isFraud ? 1 : 0, prob, risk, 0, 0, date.toISOString().replace('Z','')
        ]);
        inserted++;
    }
    stmt.finalize();
    db.run("COMMIT", err => {
        if(err) console.error("Error committing:", err);
        else console.log(`Successfully seeded ${inserted} dummy transactions using Node.js!`);
    });
  });
});

db.close();
